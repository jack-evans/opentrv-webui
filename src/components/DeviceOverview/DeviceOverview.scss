
.DeviceOverview {

  .DeviceOverview__search-bar-container {
    border-bottom: 2px solid $ui-04;
    display: flex;
    height: 100px;
    justify-content: center;
    padding: 30px 0 30px 0;

    .DeviceOverview__search-bar {
      width: 500px;
    }
  }

  .DeviceOverview__content {
    display: flex;
    flex-wrap: wrap;
    padding-bottom: 20px;
  }

  .DeviceOverview__content-loading {
    align-content: center;
    justify-content: center;
    padding-top: 10px;
    text-align: center;
  }

  .DeviceOverview__content-empty {
    align-content: center;
    flex: 1 1 auto;
    flex-flow: column;
    height: calc(100vh - 260px);
    justify-content: center;

    .DeviceOverview__content-empty-message {
      padding-bottom: 10px;
    }
  }
}