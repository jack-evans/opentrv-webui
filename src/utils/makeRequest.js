
export default function makeRequest (apiPath, options) {
  options.mode = 'cors'
  if (options.headers) {
    options.headers['Content-Type'] = 'application/json'
  } else {
    options.headers = {
      'Content-Type': 'application/json'
    }
  }

  return global.fetch(`http://localhost:3001${apiPath}`, options)
    .then(internal.checkStatus)
    .then(internal.parseJSON)
}

function checkStatus (response) {
  if (response.status >= 200 && response.status < 300) {
    return response
  } else {
    let error = new Error(response.statusText)
    error.response = response
    throw error
  }
}

function parseJSON (response) {
  return response.json()
}

export const internal = {
  checkStatus: checkStatus,
  parseJSON: parseJSON
}
