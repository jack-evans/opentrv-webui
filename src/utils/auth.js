import makeRequest from './makeRequest'
import decode from 'jwt-decode'

export function isAuthenticated () {
  const token = getToken()
  if (!token) {
    return false
  } else {
    const decodedToken = decode(token)
    const date = new Date(0)
    date.setUTCSeconds(decodedToken.exp)
    if (date < new Date()) {
      return false
    } else {
      return true
    }
  }
}

export function loginUser (user) {
  let url = '/api/v1/user/login'
  let options = {
    method: 'POST',
    json: true,
    headers: {
      'content-type': 'application/json'
    },
    body: JSON.stringify(user)
  }

  return makeRequest(url, options)
    .then(res => {
      setToken(res.token)
      return Promise.resolve()
    })
}

export function setToken (token) {
  window.localStorage.setItem('token', token)
}

export function getToken () {
  return window.localStorage.getItem('token')
}

export function logout () {
  window.localStorage.removeItem('token')
}
