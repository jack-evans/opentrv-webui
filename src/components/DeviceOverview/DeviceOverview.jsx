import React, { Component, Fragment } from 'react'
import { Button, Loading, Search } from 'carbon-components-react'
import DeviceTile from '../DeviceTile/DeviceTile'
import makeRequest from '../../utils/makeRequest'
import { getToken } from '../../utils/auth'

class DeviceOverview extends Component {
  constructor (props) {
    super(props)
    this.state = {
      devices: [],
      visibleDevices: [],
      isLoading: true
    }
    this.search = ''
    this.timer = {}

    this.discoverDevicesButtonOnClick = this.discoverDevicesButtonOnClickHandler.bind(this)
    this.searchOnChange = this.searchOnChangeHandler.bind(this)
    this.searchDevices = this.filterDevices.bind(this)
  }

  componentDidMount () {
    const tick = () => {
      return DeviceOverview.retrieveDevices()
        .then(devices => {
          this.setState({
            devices: devices,
            visibleDevices: devices.filter(this.searchDevices),
            isLoading: false
          })

          this.timer = setTimeout(tick, 1000)
        })
    }
    tick()
  }

  componentWillUnmount () {
    clearTimeout(this.timer)
  }

  searchOnChangeHandler (event) {
    this.search = event.target.value
    this.setState({visibleDevices: this.state.devices.filter(this.searchDevices)})
  }

  filterDevices (device) {
    const searchTerm = this.search.toLowerCase()

    return (
      (device.name.toLowerCase().indexOf(searchTerm) !== -1) ||
      (device.currentTemperature.toString().indexOf(searchTerm) !== -1)
    )
  }

  async discoverDevicesButtonOnClickHandler () {
    let retrievedDevices = await DeviceOverview.discoverDevices('yes')
    this.setState({
      devices: retrievedDevices,
      visibleDevices: retrievedDevices
    })
  }

  static retrieveDevices () {
    return DeviceOverview.discoverDevices('no')
  }

  static discoverDevices (userFlag) {
    let apiPath = `/api/v1/devices?user=${userFlag}`
    let options = {
      method: 'GET',
      json: true,
      headers: {
        'x-opentrv-token': getToken()
      }
    }

    return makeRequest(apiPath, options)
  }

  deleteDevice (id) {
    let apiPath = `/api/v1/devices/${id}`
    let options = {
      method: 'DELETE',
      json: true,
      headers: {
        'x-opentrv-token': getToken()
      }
    }

    return makeRequest(apiPath, options)
      .then(() => {
        return DeviceOverview.discoverDevices('no')
      })
      .then(retrievedDevices => {
        this.setState({
          devices: retrievedDevices,
          visibleDevices: retrievedDevices
        })
      })
  }

  render () {
    let deviceOverviewContent
    let deviceOverviewContentClass = 'DeviceOverview__content'

    if (this.state.isLoading) {
      deviceOverviewContentClass += ' DeviceOverview__content-loading'
      deviceOverviewContent = (
        <div>
          <div style={{'display': 'flex', 'justifyContent': 'center'}}>
            <Loading small withOverlay={false} />
          </div>
          <div>
            Loading. Please Wait...
          </div>
        </div>
      )
    } else if (this.state.devices.length < 1) {
      deviceOverviewContentClass += ' DeviceOverview__content-empty'
      deviceOverviewContent = (
        <div>
          <div className='DeviceOverview__content-empty-message'>
            <h3>You currently do not have any devices registered</h3>
          </div>
          <div>
            <Button
              className='DeviceOverview__discover-button'
              icon='add--outline'
              iconDescription='Find devices that are on your network'
              onClick={this.discoverDevicesButtonOnClick}
            >
              Discover my devices
            </Button>
          </div>
        </div>
      )
    } else {
      // This is where the tiles for each device are build up
      deviceOverviewContent = this.state.visibleDevices.map(device => {
        return (
          <DeviceTile
            key={device.id}
            id={device.id}
            name={device.name}
            temp={device.currentTemperature}
            active={device.active}
            handleDelete={() => { this.deleteDevice(device.id) }}
          />
        )
      })
    }

    return (
      <div className='DeviceOverview'>
        <div className='DeviceOverview__search-bar-container'>
          <Fragment>
            <Search
              className='DeviceOverview__search-bar'
              small
              id='device-search'
              labelText='Search'
              placeHolderText='Search Devices'
              onChange={this.searchOnChange}
            />
          </Fragment>
        </div>
        <div className={deviceOverviewContentClass}>
          {deviceOverviewContent}
        </div>
      </div>
    )
  }
}

export default DeviceOverview
