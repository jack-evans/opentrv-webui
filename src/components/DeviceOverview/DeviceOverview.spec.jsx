import React from 'react'
import DeviceOverview from './DeviceOverview'
import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { getToken } from '../../utils/auth'
jest.mock('../../utils/auth')

const nock = require('nock')

Enzyme.configure({ adapter: new Adapter() })

describe('DeviceOverview.jsx', () => {
  beforeEach(() => {
    getToken.mockReturnValue('fakeJWT')
  })

  it('renders without crashing', () => {
    expect(shallow(<DeviceOverview />)).toMatchSnapshot()
  })

  it('calls the componentDidMount function when it is created', () => {
    const componentDidMountSpy = jest.spyOn(DeviceOverview.prototype, 'componentDidMount')
    mount(<DeviceOverview />)
    expect(componentDidMountSpy).toHaveBeenCalledTimes(1)
    componentDidMountSpy.mockRestore()
  })

  it('displays a loading panel when rendered', () => {
    const wrapper = mount(<DeviceOverview />)
    expect(wrapper.find('.DeviceOverview__content-loading')).toHaveLength(1)
  })

  describe('componentDidMount', () => {
    let getNock

    beforeEach(() => {
      jest.useFakeTimers()
      getNock = nock('http://localhost:3001')
        .get('/api/v1/devices')
    })

    afterEach(() => {
      jest.useRealTimers()
      nock.cleanAll()
    })

    it('sets the devices state to an empty array when retrieve devices returns an empty array', () => {
      getNock.reply(200, [])
      const wrapper = mount(<DeviceOverview />)
      expect(wrapper.state('devices')).toEqual([])
    })

    it.skip('sets the devices state to an array of devices when retrieve devices returns devices', () => {
      getNock.reply(200, [{id: '1234'}])
      const wrapper = mount(<DeviceOverview />)

      expect(wrapper.state('devices')).toEqual([])
      return Promise.resolve()
        .then(() => {
          jest.runAllTimers()
          wrapper.update()
          let component = wrapper.instance()
          component.forceUpdate()
          expect(component.state.devices).toEqual([{id: '1234'}])
        })
    })
  })

  describe('when no devices are found on the server', () => {
    let wrapper

    beforeAll(() => {
      wrapper = mount(<DeviceOverview />)
      wrapper.setState({devices: [], visibleDevices: [], isLoading: false})
    })

    afterAll(() => {
      wrapper.unmount()
    })

    it('renders a empty message', () => {
      expect(wrapper.find('.DeviceOverview__content-empty')).toHaveLength(1)
    })

    it('renders a discover devices button', () => {
      expect(wrapper.find('.DeviceOverview__discover-button').length).toBeGreaterThan(1)
    })
  })

  describe('when devices are found on the server', () => {
    let wrapper
    let devicesFromServer = [
      {
        'id': '6f710966-c2da-4998-9bce-d4ba8971068d',
        'currentTemperature': 24.2,
        'ambientTemperature': 18,
        'name': 'living room - radiator 1',
        'serialId': 'OTRV-BXTI97EO1I',
        'active': false,
        'activeSchedules': [],
        'metadata': {}
      }, {
        'id': 'b9c23059-4e38-4e2f-b191-3c0ceec0cdcd',
        'currentTemperature': 23.3,
        'ambientTemperature': 18,
        'name': 'Device 8',
        'serialId': 'OTRV-2SHCM8CLUL',
        'active': false,
        'activeSchedules': [],
        'metadata': {}
      }, {
        'id': 'cf2a7c5e-5fc9-420a-a8ca-0e59287c4b7e',
        'currentTemperature': 31.9,
        'ambientTemperature': 18,
        'name': 'Device 2',
        'serialId': 'OTRV-5GZ3VMLVY2',
        'active': false,
        'activeSchedules': [],
        'metadata': {}
      }
    ]

    beforeEach(() => {
      wrapper = mount(<DeviceOverview />)
      wrapper.setState({devices: devicesFromServer, visibleDevices: devicesFromServer, isLoading: false})
    })

    afterEach(() => {
      wrapper.unmount()
    })

    it('renders a matching number of DeviceTiles', () => {
      expect(wrapper.find('.DeviceTile')).toHaveLength(devicesFromServer.length)
    })
  })

  describe('when a user searches through their devices', () => {
    let wrapper
    let devicesFromServer = [
      {
        'id': '6f710966-c2da-4998-9bce-d4ba8971068d',
        'currentTemperature': 24.2,
        'ambientTemperature': 18,
        'name': 'living room - radiator 1',
        'serialId': 'OTRV-BXTI97EO1I',
        'active': false,
        'activeSchedules': [],
        'metadata': {}
      }, {
        'id': 'b9c23059-4e38-4e2f-b191-3c0ceec0cdcd',
        'currentTemperature': 23.3,
        'ambientTemperature': 18,
        'name': 'Device 8',
        'serialId': 'OTRV-2SHCM8CLUL',
        'active': false,
        'activeSchedules': [],
        'metadata': {}
      }, {
        'id': 'cf2a7c5e-5fc9-420a-a8ca-0e59287c4b7e',
        'currentTemperature': 31.9,
        'ambientTemperature': 18,
        'name': 'Device 2',
        'serialId': 'OTRV-5GZ3VMLVY2',
        'active': false,
        'activeSchedules': [],
        'metadata': {}
      }
    ]

    beforeEach(() => {
      wrapper = mount(<DeviceOverview />)
      wrapper.setState({devices: devicesFromServer, visibleDevices: devicesFromServer, isLoading: false})
      wrapper.find('input').simulate('change', { target: { value: 'device' } })
    })

    afterEach(() => {
      wrapper.unmount()
    })

    it('sets visible devices to the 2 valid devices', () => {
      expect(wrapper.state('visibleDevices').length).toEqual(2)
    })

    it('displays only 2 of the 3 DeviceTiles', () => {
      expect(wrapper.find('.DeviceTile')).toHaveLength(2)
    })
  })

  describe('when a user presses delete on a DeviceTile', () => {
    let wrapper
    let deleteDeviceSpy
    let devicesFromServer = [
      {
        'id': '6f710966-c2da-4998-9bce-d4ba8971068d',
        'currentTemperature': 24.2,
        'ambientTemperature': 18,
        'name': 'living room - radiator 1',
        'serialId': 'OTRV-BXTI97EO1I',
        'active': false,
        'activeSchedules': [],
        'metadata': {}
      }
    ]

    beforeEach(() => {
      wrapper = shallow(<DeviceOverview />)
      wrapper.setState({devices: devicesFromServer, visibleDevices: devicesFromServer, isLoading: false})
      deleteDeviceSpy = jest.spyOn(DeviceOverview.prototype, 'deleteDevice')

      wrapper.find('DeviceTile').prop('handleDelete')()
    })

    afterEach(() => {
      wrapper.unmount()
      deleteDeviceSpy.mockReset()
    })

    it('calls the deleteDevice function', () => {
      expect(deleteDeviceSpy).toHaveBeenCalledTimes(1)
    })

    it('calls the deleteDevice function with the id of the Device', () => {
      expect(deleteDeviceSpy).toHaveBeenCalledWith(devicesFromServer[0].id)
    })
  })
})
