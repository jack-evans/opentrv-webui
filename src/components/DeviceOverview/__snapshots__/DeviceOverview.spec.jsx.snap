// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceOverview.jsx renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DeviceOverview />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="DeviceOverview__search-bar-container"
        >
          <UNDEFINED>
            <Search
              className="DeviceOverview__search-bar"
              id="device-search"
              labelText="Search"
              onChange={[Function]}
              placeHolderText="Search Devices"
              small={true}
              type="text"
            />
          </UNDEFINED>
        </div>,
        <div
          className="DeviceOverview__content DeviceOverview__content-loading"
        >
          <div>
            <div
              style={
                Object {
                  "display": "flex",
                  "justifyContent": "center",
                }
              }
            >
              <Loading
                active={true}
                small={true}
                withOverlay={false}
              />
            </div>
            <div>
              Loading. Please Wait...
            </div>
          </div>
        </div>,
      ],
      "className": "DeviceOverview",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <UNDEFINED>
            <Search
              className="DeviceOverview__search-bar"
              id="device-search"
              labelText="Search"
              onChange={[Function]}
              placeHolderText="Search Devices"
              small={true}
              type="text"
            />
          </UNDEFINED>,
          "className": "DeviceOverview__search-bar-container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Search
              className="DeviceOverview__search-bar"
              id="device-search"
              labelText="Search"
              onChange={[Function]}
              placeHolderText="Search Devices"
              small={true}
              type="text"
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "className": "DeviceOverview__search-bar",
              "id": "device-search",
              "labelText": "Search",
              "onChange": [Function],
              "placeHolderText": "Search Devices",
              "small": true,
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": Symbol(react.fragment),
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div>
            <div
              style={
                Object {
                  "display": "flex",
                  "justifyContent": "center",
                }
              }
            >
              <Loading
                active={true}
                small={true}
                withOverlay={false}
              />
            </div>
            <div>
              Loading. Please Wait...
            </div>
          </div>,
          "className": "DeviceOverview__content DeviceOverview__content-loading",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                style={
                  Object {
                    "display": "flex",
                    "justifyContent": "center",
                  }
                }
              >
                <Loading
                  active={true}
                  small={true}
                  withOverlay={false}
                />
              </div>,
              <div>
                Loading. Please Wait...
              </div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Loading
                  active={true}
                  small={true}
                  withOverlay={false}
                />,
                "style": Object {
                  "display": "flex",
                  "justifyContent": "center",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "active": true,
                  "small": true,
                  "withOverlay": false,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Loading. Please Wait...",
              },
              "ref": null,
              "rendered": "Loading. Please Wait...",
              "type": "div",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="DeviceOverview__search-bar-container"
          >
            <UNDEFINED>
              <Search
                className="DeviceOverview__search-bar"
                id="device-search"
                labelText="Search"
                onChange={[Function]}
                placeHolderText="Search Devices"
                small={true}
                type="text"
              />
            </UNDEFINED>
          </div>,
          <div
            className="DeviceOverview__content DeviceOverview__content-loading"
          >
            <div>
              <div
                style={
                  Object {
                    "display": "flex",
                    "justifyContent": "center",
                  }
                }
              >
                <Loading
                  active={true}
                  small={true}
                  withOverlay={false}
                />
              </div>
              <div>
                Loading. Please Wait...
              </div>
            </div>
          </div>,
        ],
        "className": "DeviceOverview",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <UNDEFINED>
              <Search
                className="DeviceOverview__search-bar"
                id="device-search"
                labelText="Search"
                onChange={[Function]}
                placeHolderText="Search Devices"
                small={true}
                type="text"
              />
            </UNDEFINED>,
            "className": "DeviceOverview__search-bar-container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Search
                className="DeviceOverview__search-bar"
                id="device-search"
                labelText="Search"
                onChange={[Function]}
                placeHolderText="Search Devices"
                small={true}
                type="text"
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "className": "DeviceOverview__search-bar",
                "id": "device-search",
                "labelText": "Search",
                "onChange": [Function],
                "placeHolderText": "Search Devices",
                "small": true,
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": Symbol(react.fragment),
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div>
              <div
                style={
                  Object {
                    "display": "flex",
                    "justifyContent": "center",
                  }
                }
              >
                <Loading
                  active={true}
                  small={true}
                  withOverlay={false}
                />
              </div>
              <div>
                Loading. Please Wait...
              </div>
            </div>,
            "className": "DeviceOverview__content DeviceOverview__content-loading",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  style={
                    Object {
                      "display": "flex",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Loading
                    active={true}
                    small={true}
                    withOverlay={false}
                  />
                </div>,
                <div>
                  Loading. Please Wait...
                </div>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <Loading
                    active={true}
                    small={true}
                    withOverlay={false}
                  />,
                  "style": Object {
                    "display": "flex",
                    "justifyContent": "center",
                  },
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "active": true,
                    "small": true,
                    "withOverlay": false,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Loading. Please Wait...",
                },
                "ref": null,
                "rendered": "Loading. Please Wait...",
                "type": "div",
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
